File: samples-java/src/main/java/io/github/ilaborie/loops/samples/recursion/TransformTR.java
package io.github.ilaborie.loops.samples.recursion;

import io.github.ilaborie.loops.samples.Element;

import java.util.ArrayList;
import java.util.List;

import static io.github.ilaborie.loops.samples.Elements.getSomeElements;

public class TransformTR {

    public static void main(String[] args) {
        List<Element> elements = getSomeElements();
        List<String> result = transform(elements);
        result.forEach(System.out::println);
    }

    public static List<String> transform(List<Element> elements) {
        return aux(elements, new ArrayList<>());
    }

    private static List<String> aux(List<Element> elements, List<String> acc) {
        if (elements.isEmpty()) {
            return acc;
        }

        Element head = elements.get(0);
        List<Element> tail = elements.subList(1, elements.size());

        String transformed = head.getName() + " \\o/";
        acc.add(transformed);

        return aux(tail, acc);
    }

}

File: samples-kotlin/src/main/kotlin/io/github/ilaborie/loops/samples/recursion/transformTR.kt
package io.github.ilaborie.loops.samples.recursion

import io.github.ilaborie.loops.samples.ElementKt
import io.github.ilaborie.loops.samples.getSomeElementsKt


fun main() {
    val elements = getSomeElementsKt()
    val result = transformTR(elements)
    result.forEach { println(it) }
}

fun transformTR(elements: List<ElementKt>): List<String> {
    tailrec fun aux(elements: List<ElementKt>, acc: List<String>): List<String> {
        if (elements.isEmpty()) {
            return acc
        }

        val head = elements.first()
        val tail = elements.drop(1)

        val transformed = head.name + " \\o/"
        return aux(tail, acc + listOf(transformed))
    }

    return aux(elements, emptyList())
}

fun transformTRAlt(elements: List<ElementKt>): List<String> {
    tailrec fun aux(elements: List<ElementKt>, acc: List<String>): List<String> =
        if (elements.isEmpty()) acc
        else aux(elements.drop(1), acc + (elements.first().name + " \\o/"))

    return aux(elements, emptyList())
}